src/invoke_training/lora/layers/lora_conv_layer.py

```diff
@@ -0,0 +1,131 @@
+        raise NotImplementedError(
+            "LoRAConvLayer cannot be used directly. Use LoRAConv1dLayer, LoRAConv2dLayer, or LoRAConv3dLayer instead."
+        )
+
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        kernel_size: typing.Union[int, tuple[int]] = 1,
+        stride: typing.Union[int, tuple[int]] = 1,
+        padding: typing.Union[str, int, tuple[int]] = 0,
+        rank: int = 4,
+        alpha: float = 1.0,
+        device: torch.device = None,
+        dtype: torch.dtype = None,
+    ):
+        """Initialize a LoRAConvLayer.
+        Args:
+            in_channels (int): The number of channels expected on inputs to this layer.
+            out_channels (int): The number of channels on outputs from this layer.
+            kernel_size: The kernel_size of the conv layer that this LoRA layer is mirroring. See torch.nn.Conv* docs.
+            stride: The stride of the conv layer that this LoRA layer is mirroring. See torch.nn.Conv* docs.
+            padding: The padding of the conv layer that this LoRA layer is mirroring. See torch.nn.Conv* docs.
+            rank (int, optional): The internal rank of the layer. See the paper for details.
+            alpha (float, optional): A scaling factor that enables tuning the rank without having to adjust the learning
+                rate. The recommendation from the paper is to set alpha equal to the first rank that you try and then do
+                not tune it further. See the paper for more details.
+            device (torch.device, optional): Device where weights will be initialized.
+            dtype (torch.dtype, optional): Weight dtype.
+        Raises:
+            ValueError: If the rank is greater than either in_channels or out_channels.
+        """
+        super().__init__()
+
+        if rank > min(in_channels, out_channels):
+            raise ValueError(f"LoRA rank {rank} must be less than or equal to {min(in_channels, out_channels)}")
+
+        self._down = self.conv_module(
+            in_channels,
+            rank,
+            kernel_size=kernel_size,
+            stride=stride,
+            padding=padding,
+            bias=False,
+            device=device,
+            dtype=dtype,
+        )
+        self._up = self.conv_module(rank, out_channels, kernel_size=1, stride=1, bias=False, device=device, dtype=dtype)
+
+        # Register alpha as a buffer so that it is not trained, but still gets saved to the state_dict.
+        self.register_buffer("alpha", torch.tensor(alpha, device=device, dtype=dtype))
+
+        self._rank = rank
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        # This initialization is based on Microsoft's implementation:
+        # https://github.com/microsoft/LoRA/blob/998cfe4d351f4d6b4a47f0921dec2397aa0b9dfe/loralib/layers.py#L279
+        torch.nn.init.kaiming_uniform_(self._down.weight, a=math.sqrt(5))
```

Comment:
What does this do?

Reply:
I'm not exactly sure what you mean by *this*, but I'll try to explain.

`reset_parameters()` sets the initial LoRA weights before they are trained.

Different weight initializations have different convergence properties. I don't know the details of the `kaiming_uniform_` initialization, but generally the goal is to initialize the weights such that when you stack a bunch of layers in a network you don't exponentially grow/diminish the activations in the forward pass (or the gradients in the backwards pass).

I copied this `kaiming_uniform_` configuration from Microsoft's LoRA implementation, and it looks like they copied it from the [default torch.nn.Linear initialization](https://github.com/pytorch/pytorch/blob/2a138d7f1df70d7b61bdc59fd43d42f8f59a4262/torch/nn/modules/linear.py#L104-L107). (If you really want to go down a rabbit hole, follow that think for the explanation of why torch uses it.)

For LoRA layers, the 'up' weights are initialized to 0. This is done so that the LoRA layer can be added to a network and it will initially have no effect on the network behaviour. Then, as it is trained, it will gradually exert more influence over the base model.

Reply:
Gotcha thanks! I was mostly asking about the `kaiming_uniform_` function, and the reason why it's run with `a=math.sqrt(5)`

I went a little bit down the rabbithole and it seemed like it'd take me a while to understand it fully so I was curious if you had an answer quickly for it. Not really important as long as it works and is documented to be useful. That explanation for what the initialization for is plenty for me, thanks!